main/src/ca/uwaterloo/flix/language/phase/unification/BoolMinimization.scala

```diff
@@ -0,0 +1,482 @@
+    }
+
+    @tailrec
+    def aux(terms0: List[Formula], bindings: Map[Formula.Var, Formula], acc: List[Formula]): List[Formula] = terms0 match {
+      case Nil =>
+        acc.reverse
+      case (v: Var) :: rest if bindings.contains(v) =>
+        aux(subst(v, bindings) :: rest, bindings, acc)
+      case (v: Var) :: rest =>
+        aux(rest, bindings + (v -> True), v :: acc)
+      case Not(v: Var) :: rest if bindings.contains(v) =>
+        aux(subst(Not(v), bindings) :: rest, bindings, acc)
+      case Not(v: Var) :: rest =>
+        aux(rest, bindings + (v -> False), Not(v) :: acc)
+      case f :: rest if bindings.nonEmpty =>
+        val replacedF = subst(f, bindings)
+        replacedF match {
+          case True =>
+            aux(rest, bindings, acc)
+          case False =>
+            // false shouldn't occur here so shortcutting doesn't matter
+            aux(rest, bindings, False :: acc)
+          case v: Var if bindings.contains(v) =>
+            aux(subst(v, bindings) :: rest, bindings, acc)
+          case v: Var =>
+            aux(rest, bindings + (v -> True), v :: acc)
+          case Not(v: Var) =>
+            aux(rest, bindings + (v -> False), Not(v) :: acc)
+          case Not(_) | And(_) | Or(_) =>
+            aux(rest, bindings, replacedF :: acc)
+        }
+      case f :: rest => aux(rest, bindings, f :: acc)
+    }
+
+    val propagatedTerms = aux(terms, Map.empty, Nil)
+    // This last normalization is for sorting
+    normalizeTerms(propagatedTerms)
+  }
+
+  /**
+    * Build `Not(f)` and performs shallow minimization.
+    */
+  def mkNot(f: Formula): Formula = {
+    import Formula._
+    f match {
+      case True => False
+      case False => True
+      case Var(v) => Not(Var(v))
+      case Not(term) => term
+      case And(terms) => Not(And(terms))
+      case Or(terms) => Not(Or(terms))
+    }
+  }
+
+  /**
+    * Merges the conjunctions if arguments are conjunctions themselves and
+    * performs shallow minimization along with unit propagation.
+    */
+  def mkAnd(f: Formula*): Formula = {
+    mkAnd(f.toList)
+  }
+
+  /**
+    * Merges the conjunctions in `f` into the new conjunction and performs
+    * shallow minimization.
+    */
+  def mkAnd(f: List[Formula]): Formula = {
+    import Formula._
+    val baseValue: Option[List[Formula]] = Some(Nil)
+    val hoistedTerms = f.foldRight(baseValue) {
+      case (term, opt) => opt.flatMap(
+        acc => term match {
+          case True => Some(acc)
+          case False => None
+          case v: Var => Some(v :: acc)
+          case n: Not => Some(n :: acc)
+          case And(terms) => Some(terms ++ acc)
+          case or: Or => Some(or :: acc)
+        }
+      )
+    }
+    val finalTerms = hoistedTerms
+      .flatMap(normalizeTerms)
+      .flatMap(unitPropagation)
+    finalTerms match {
+      case None => False
+      case Some(term :: Nil) => term
+      case Some(terms) => And(terms)
+    }
```

Comment:
Using `++` and `flatMap` is usually fine, but since this is a performance critical place, maybe we want something better?